### ISSUE TITLE
Spatial Transformer Networks cuts off last row and column off input

### ISSUE THREAD

**Issue by hughbzhang** [tensorflow/models#2016]
Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

Also, please understand that many of the models included in this repository are experimental and research-style code. If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **What is the top-level directory of the model you are using**:
models/transformer
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Only the script attached below to reproduce the issue
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Tested on both MacOS Sierra and Linux Ubuntu 16.04.2 LTS
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
1.2.1
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
n/a
- **GPU model and memory**:
n/a
- **Exact command to reproduce**:

```python
import tensorflow as tf
import numpy as np
from spatial_transformer import transformer

def identity():

    original_theta = np.array([[1., 0., 0.],
                               [0., 1., 0.]])

    theta = tf.Variable(initial_value=np.expand_dims(original_theta.flatten(), 0))
    
    original_U = np.reshape(np.arange(25, dtype="float32"), [5,5])
    reshaped_U = np.expand_dims(np.expand_dims(original_U, 2), 0)
    U = tf.Variable(initial_value=reshaped_U)

    assert(U.shape == (1,5,5,1))
    assert(theta.shape == (1,6))

    raw_output = transformer(U, theta, (5,5))
    
    sess = tf.Session()
    sess.run(tf.global_variables_initializer())
    output = np.array(sess.run(raw_output))
   
    assert(reshaped_U.shape == (1,5,5,1))
    assert(output.shape == (1,5,5,1))

    print(output)
    print(reshaped_U)

    assert((output == reshaped_U).all())
```

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

I cloned the models/transformer directory to generalize 2D spatial transformers to 3D. In the process, I found a few bugs that caused spatial transformers to not output the exact transformation specified. After investigating, I found that there are some indexing errors which cause the problem shown above. I'd be happy to make a PR for any or all of the following if it would be helpful:
1) The fixed 2D spatial transformer
2) Newly implemented 3D spatial transformer
3) Basic test cases for both of the above

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by xcyan** [comment]
@hughbzhang perspective 3D transformer can be found here as well: https://github.com/tensorflow/models/blob/master/ptn/nets/perspective_transform.py

**Comment by xcyan** [comment]
#1320 This might also be related. @hughbzhang

**Comment by hughbzhang** [comment]
Ah, darn! Could have saved me the time to implement it myself :D. That being said, I think my original issue still applies. The above test code shows that applying the identity transformation matrix does not apply the identity transformation.

**Comment by hughbzhang** [comment]
Also, on #1320, if a pixel falls outside of the 2D grid, what is a better alternative to using the boundary pixels to calculate its value?

**Comment by xcyan** [comment]
I just curious about the fix you had, what was the issue of the spatial transformer implementation? 
@cbfinn Did you observe this issue before for your video prediction model?

**Comment by cbfinn** [comment]
I didn't observe any issues with the implementation. Though, generally, it is sometimes hard to get a model to train with it (because the gradients aren't smooth).

**Comment by hughbzhang** [comment]
Running the above code produces the following output:

```
(Output)
[[[[  0.  ]
   [  1.25]
   [  2.5 ]
   [  3.75]
   [  0.  ]]

  [[  6.25]
   [  7.5 ]
   [  8.75]
   [ 10.  ]
   [  0.  ]]

  [[ 12.5 ]
   [ 13.75]
   [ 15.  ]
   [ 16.25]
   [  0.  ]]

  [[ 18.75]
   [ 20.  ]
   [ 21.25]
   [ 22.5 ]
   [  0.  ]]

  [[  0.  ]
   [  0.  ]
   [  0.  ]
   [  0.  ]
   [  0.  ]]]]

(Target)
[[[[  0.]
   [  1.]
   [  2.]
   [  3.]
   [  4.]]

  [[  5.]
   [  6.]
   [  7.]
   [  8.]
   [  9.]]

  [[ 10.]
   [ 11.]
   [ 12.]
   [ 13.]
   [ 14.]]

  [[ 15.]
   [ 16.]
   [ 17.]
   [ 18.]
   [ 19.]]

  [[ 20.]
   [ 21.]
   [ 22.]
   [ 23.]
   [ 24.]]]]```


There are two issues. One is that the indexing causes the last row and last column to be dropped. The second is that the indices are scaled wrong (from 0 to width or height instead of 0 to width-1,height-1).

In terms of why a network using this / derivations of this code might still work, the effect of this transformation is instead of having a NxM output size, you essentially have a (N-1)x(M-1) output size. The distortion bug also merely causes there to be a different identity matrix instead of the typical eye matrix. Thus, the transformation module still "works" despite being buggy, but it is extremely confusing when trying to use it for a new project.

I'd be happy to open a PR with the fixes. Please let me know if there is anything else I can do! Running the code pasted in the original issue will replicate the bug. Thanks!

**Comment by xcyan** [comment]
Can you clarify whether the bug is in the models/transformer or somewhere else? This seems to be a serious bug. @panyx0718 @martinwicke

**Comment by hughbzhang** [comment]
@xcyan The bug is in the _interpolate function of models/transformer/spatial_transformer.py.

**Comment by xcyan** [comment]
@hughbzhang I guess you are talking about this line: https://github.com/tensorflow/models/blob/master/transformer/spatial_transformer.py#L80

Should be something like this
`x = (x + 1.0)*(width_f - 1) / 2.0`

**Comment by hughbzhang** [comment]
@xcyan Yes, that will solve the issue where the identity function produces decimals. To solve, the edges getting clipped, you also need something like:

```
            x1 = tf.clip_by_value(x1, zero, max_x+1)
            x1_clipped = tf.clip_by_value(x1, zero, max_x)
            y1 = tf.clip_by_value(y1, zero, max_y+1)
            y1_clipped = tf.clip_by_value(y1, zero, max_y)
            ......... other code ..............
            base_y0 = base + y0*dim2
            base_y1 = base + y1_clipped*dim2
            idx_a = base_y0 + x0
            idx_b = base_y1 + x0
            idx_c = base_y0 + x1_clipped
            idx_d = base_y1 + x1_clipped
```

I'd be happy to update the fix as well as the other test cases I wrote to sanity check myself.

**Comment by xcyan** [comment]
Thank you for clarification. Please be patient since I would not be able to review your PR (I am not a contributor of 2D transformer). Instead, we will fix the related problem in [perspective 3D transformers](https://github.com/tensorflow/models/tree/master/ptn). @arkanath

**Comment by hughbzhang** [comment]
Sounds good! Glad to be of help. Side note, I also noticed that tensorflow now has tf.meshgrid while doing my implementation. I didn't find any bugs in the original implementation, but might be easier to use!

**Comment by xcyan** [comment]
@aselle Thanks! To fix this bug, do you think it's a good idea for @hughbzhang to submit a PR?

**Comment by hughbzhang** [comment]
@xcyan @aselle Checking in on this again!

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
